transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java

```diff
@@ -82,15 +81,58 @@ public WebSocketExtensionSpi negotiate(ExtensionHeader extension, WsResourceAddr
      * @return object representing the list of negotiated  WebSocketExtensionSpi instances
      * @throws ProtocolException
      */
-    public ActiveExtensions negotiateWebSocketExtensions(ResourceAddress address,
-                                                         HttpAcceptSession session,
-                                                         String headerName,
-                                                         List<String> clientRequestedExtensions) throws ProtocolException {
-        // TODO: implement this method, based loosely on WsExtensionUtils.negotiateWebSocketExtensions (it will replace
-        // that method). Note that (as commented in that method) we do not need to have any notion of mandatory extensions 
-        // the client must negotiate.
-        throw new UnsupportedOperationException ();
-        //return ActiveExtensions.EMPTY;
+    public ActiveExtensions negotiateWebSocketExtensions(WsResourceAddress address, HttpAcceptSession session,
+        String headerName, List<String> clientRequestedExtensions) throws ProtocolException {
+
+        ActiveExtensions result = ActiveExtensions.EMPTY;
+        if (clientRequestedExtensions != null) {
+            // Maybe come back and cache supportedExtensionHeaders ??, but need more info on when they are loaded to
+            // decide
+            List<ExtensionHeader> supportedExtensionHeaders = toWsExtensions(this.getExtensionNames());
+            List<ExtensionHeader> requestedExtensions = toWsExtensions(clientRequestedExtensions);
+
+            // Since the client may have provided parameters in their
+            // extensions, we retain the common requested extensions, not
+            // the common supported extensions as was previously done.
+            requestedExtensions.retainAll(supportedExtensionHeaders);
+
+            // get accepted extensions
+            List<ExtensionHeader> acceptedExtensions = new ArrayList<>(requestedExtensions.size());
+            // negotiated Extensions will be used when ActiveExtensions is changed to new API, so leaving it in
+            // List<WebSocketExtensionSpi> negotiatedExtensions = new ArrayList<>();
+            for (ExtensionHeader candidate : requestedExtensions) {
+                WebSocketExtensionFactorySpi extension = factoriesRO.get(candidate.getExtensionToken());
+                WebSocketExtensionSpi negotiatedExtension = extension.negotiate(candidate, address);
+                // negotiated can be null if the extension doesn't want to be envolved
+                if (negotiatedExtension != null) {
+                    acceptedExtensions.add(candidate);
+                    // negotiatedExtensions.add(negotiatedExtension);
+                }
+            }
+            // TODO: This is where we can add server-initiated coordinated extension parameters.
```

Comment:
Please remove this comment, I have no idea what it means! so there is no point carrying it over from 4.0